### PR TITLE
feat: add graceful shutdown

### PR DESCRIPTION
**Pull request by zmstone**
background:
EMQ X node is booted with boot file embedded mode.
the applications boot order is generated from rebar.confg + top-sorted order done by rebar/relx
`emqx` app is instructed to be only loaded at boot,
`emqx_machine` starts `emqx` app after some init work is done.
In order to gain better control over the boot order and the ability to debug boot sequence,
apps in this project are top-sorted and started by `emqx_machine`
(although they might be started in boot file too).

prior to this change, `emqx` node shutdown is done by `init:stop()`
which has two issues:
1. might have undesired stop order of the applications (different than customised order defined)
2. `init:stop()` returns `ok` before beam is terminated, which leads to a `kill -s 0` looping check in the bash script to wait for beam to exit --- which is no longer necessary after this change.

in this change, emqx_machine_terminator is added to stop apps
in defined order and then terminate the node in infinite loop

### CODE REVIEW

File: apps/emqx_machine/src/emqx_machine_signal_handler.erl [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    emqx_machine_terminator:graceful(),
+    {ok, State};
+handle_event(Event, State) ->
+    %% delegate other events back to erl_signal_handler
+    %% erl_signal_handler does not make use of the State
+    %% so we can pass whatever from here
+    erl_signal_handler:handle_event(Event, State),
+    {ok, State}.
+
+handle_info(stop, State) ->
```
Comment (zhongwencool) [inline apps/emqx_machine/src/emqx_machine_signal_handler.erl:48]: why don't ignore other messages like `erl_signal_server`?
```
handle_info(_Info, S) ->
    {ok, S}.
```

File: apps/emqx_machine/src/emqx_machine.erl [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+    application:stop(App).
+
+ensure_apps_started() ->
+    lists:foreach(fun start_one_app/1, sorted_reboot_apps()).
+
+start_one_app(App) ->
+    ?SLOG(debug, #{msg => "starting_app", app => App}),
+    case application:ensure_all_started(App) of
+        {ok, Apps} ->
+            ?SLOG(debug, #{msg => "started_apps", apps => [App | Apps]});
```
Comment (zhongwencool) [inline apps/emqx_machine/src/emqx_machine.erl:113]: [App | lists:delete(App,Apps)].

File: apps/emqx_machine/src/emqx_machine_terminator.erl [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+            %% Erlang VM goes down before this receive.
+            %% In which case, the remote caller will get {badrpc, nodedown}
+            receive {'DOWN', Ref, process, Pid, _} -> ok end
+    end.
+
+init(_) ->
+    ok = emqx_machine_signal_handler:start(),
+    {ok, #{}}.
+
+handle_info(?DO_IT, State) ->
```
Comment (zhongwencool) [inline apps/emqx_machine/src/emqx_machine_terminator.erl:64]: write some log:
```erlang
try
        ok = emqx_machine:stop_apps(normal)
    catch Type:Reason ->
        %% maybe also log to structured error log?
        ?ULOG("~s stop error: ~p~n", [emqx_app:get_release(), {Type, Reason}]),
        error({Type, Reason})
    after
        try
            case application:which_applications() of
                [] -> ok;
                Apps ->
                    ?LOG(error, "Halting Erlang VM with applications:~p~n  ", [element(1, A)||A<- Apps]),
                    %% console where foreground was running (or systemd journal) will
                    %% contain information about graceful termination.
                    io:format(standard_error, "Gracefully halting Erlang VM~n", [])
            end
        after
            init:stop()
        end
    end.
```

**Comment by zhongwencool** [comment]
[SIGHUP](https://www.wikiwand.com/en/SIGHUP#:~:text=On%20POSIX%2Dcompliant%20platforms%2C%20SIGHUP,in%20the%20header%20file%20signal.) applications that did not require a controlling terminal, such as daemons, would re-purpose SIGHUP as a signal to re-read configuration files, or reinitialize. This convention survives to this day in packages such as Apache and Sendmail. We don't support any of those two cases yet, so ignore it.

[SIGTSTP](https://en-academic.com/dic.nsf/enwiki/2219096) is sent to a process by its controlling terminal when the user presses the special SUSP key combination (usually control-Z). By default, SIGTSTP causes the process receiving it to stop until a SIGCONT signal is received. Unlike the similar SIGSTOP signal, a process can register a signal handler for or ignore SIGTSTP.. we don't handle SIGCONT, it makes a SIGTSTP handler less useful here， so ignore it.

http://erlang.org/doc/man/os.html#set_signal-2
```erlang
os:set_signal(sighup,ignore),
os:set_signal(sighstp,ignore),
os:set_signal(sigterm,handle). %% note default is handle
```

File: apps/emqx_machine/src/emqx_machine.erl (hunk 2 above)
Comment (zmstone) [reply apps/emqx_machine/src/emqx_machine.erl:113]: Ah, I thought this list only includes dependent apps.
it actually includes the app itself, will remove `[App |`

**Comment by zmstone** [comment]
> [SIGHUP](https://www.wikiwand.com/en/SIGHUP#:~:text=On%20POSIX%2Dcompliant%20platforms%2C%20SIGHUP,in%20the%20header%20file%20signal.) applications that did not require a controlling terminal, such as daemons, would re-purpose SIGHUP as a signal to re-read configuration files, or reinitialize. This convention survives to this day in packages such as Apache and Sendmail. We don't support any of those two cases yet, so ignore it.
> 
> [SIGTSTP](https://en-academic.com/dic.nsf/enwiki/2219096) is sent to a process by its controlling terminal when the user presses the special SUSP key combination (usually control-Z). By default, SIGTSTP causes the process receiving it to stop until a SIGCONT signal is received. Unlike the similar SIGSTOP signal, a process can register a signal handler for or ignore SIGTSTP.. we don't handle SIGCONT, it makes a SIGTSTP handler less useful here， so ignore it.
> 
> http://erlang.org/doc/man/os.html#set_signal-2
> 
> ```erlang
> os:set_signal(sighup,ignore),
> os:set_signal(sighstp,ignore),
> os:set_signal(sigterm,handle). %% note default is handle
> ```
Thank you.
`sigtstp` can be handy when running in interactive mode (`console`), inclined to keep.

File: apps/emqx_machine/src/emqx_machine_terminator.erl (hunk 3 above)
Comment (zmstone) [reply apps/emqx_machine/src/emqx_machine_terminator.erl:64]: the `try`: I'm not entirely sure if `_ = application:stop()` is ever going to crash? it might hang indefinitely, but not crash.

the `after` clause: `emqx_machine:stop_apps/1` only stops the ones we want to control, a lot others will remain (especially the ones come from otp), meaning it'll always log an `error` level message ?

Comment (zmstone) [reply apps/emqx_machine/src/emqx_machine_terminator.erl:64]: pushed a fix which added `try` catch in `emqx_machine:stop_one_app`

Comment (zhongwencool) [reply apps/emqx_machine/src/emqx_machine_terminator.erl:64]: `emqx_machine:stop_apps/1` maybe change in the future, The first try just make sure no matter what changes, it always run `init:stop()`.
the second try just make sure LOG don't crash, do his best to write some logs.
application:which_applications() alway not empty, write  application info to error.log not to stand_error.

Comment (zmstone) [reply apps/emqx_machine/src/emqx_machine_terminator.erl:64]: good point.
I'll add the error log to a `catch` clause, but not `after`.

File: apps/emqx_machine/src/emqx_machine_terminator.erl [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+start() ->
+    {ok, _} = gen_server:start_link({local, ?TERMINATOR}, ?MODULE, [], []),
+    %% NOTE: Do not link this process under any supervision tree
+    ok.
+
+is_running() -> is_pid(whereis(?TERMINATOR)).
+
+%% @doc Send a signal to activate the terminator.
+graceful() ->
+    ?TERMINATOR ! ?DO_IT,
```
Comment (qzhuyan) [inline apps/emqx_machine/src/emqx_machine_terminator.erl:48]: handle the case where TERMINATOR isn't alive?

Comment (qzhuyan) [reply apps/emqx_machine/src/emqx_machine_terminator.erl:48]: I think we need some confirmation here, otherwise init:stop/0 will never get called?

File: apps/emqx_machine/src/emqx_machine_signal_handler.erl [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+handle_call(_Request, State) ->
+    {ok, ok, State}.
+
+format_status(_Opt, [_Pdict,_S]) ->
+    ok.
+
+code_change(_OldVsn, State, _Extra) ->
+    {ok, State}.
+
+terminate(_Args, _State) ->
```
Comment (qzhuyan) [inline apps/emqx_machine/src/emqx_machine_signal_handler.erl:62]: shall we return the signal handling ownership back to erl_signal_handler?

File: apps/emqx_machine/src/emqx_machine_terminator.erl (hunk 4 above)
Comment (zmstone) [reply apps/emqx_machine/src/emqx_machine_terminator.erl]: changed to `gen_server:call`, and `catch` exceptions.

File: apps/emqx_machine/src/emqx_machine_signal_handler.erl (hunk 5 above)
Comment (zmstone) [reply apps/emqx_machine/src/emqx_machine_signal_handler.erl:62]: `erl_signal_handler` is just a handler module, there is only one process which is `erl_signal_server`
so nothing really to hand-back here.

File: apps/emqx_machine/src/emqx_machine.erl [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+        , ensure_apps_started/0
+        ]).
+
+-export([sorted_reboot_apps/0]).
+
+-include_lib("emqx/include/logger.hrl").
+
+%% @doc EMQ X boot entrypoint.
+start() ->
+    os:set_signal(sighup, ignore),
```
Comment (qzhuyan) [inline apps/emqx_machine/src/emqx_machine.erl:34]: ignore sighup is an unrelated change or it is required?

File: apps/emqx_machine/src/emqx_machine_signal_handler.erl [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    ok = gen_event:swap_sup_handler(
+           erl_signal_server,
+           {erl_signal_handler, []},
+           {?MODULE, []}).
+
+init({[], _}) -> {ok, #{}}.
+
+handle_event(sigterm, State) ->
+    ?ULOG("Received terminate signal, shutting down now~n", []),
+    emqx_machine_terminator:graceful(),
```
Comment (zhongwencool) [inline apps/emqx_machine/src/emqx_machine_signal_handler.erl:39]: handle early kill sigterm.
```erlang
case emqx_machine:is_ready() of 
  true -> emqx_machine_terminator:graceful();
  false -> init:stop()
end.
```

File: apps/emqx_machine/src/emqx_machine.erl [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+    ok = load_config_files(),
+
+    ok = ensure_apps_started(),
+
+    _ = emqx_plugins:load(),
+
+    ok = print_vsn(),
+
+    ok = start_autocluster(),
+    %% NOTE: keep this to the end
```
Comment (zhongwencool) [inline apps/emqx_machine/src/emqx_machine.erl:51]: In my opinion， 
1. If we want to keep this to the end, we should move this call out of `emqx_machine:start/0`.
```erlang
emqx_machine:start(),
emqx_machine_terminator:start().
```
2. I find emqx_machine_terminator link to application control, not under emqx_machine_sup, any reason for this?
I think emqx_machine_sup should monitor it by one_for_one, restart  => transient.

File: apps/emqx_machine/src/emqx_machine.erl (hunk 6 above)
Comment (zmstone) [reply apps/emqx_machine/src/emqx_machine.erl:34]: it's unrelated to graceful shutdown

File: apps/emqx_machine/src/emqx_machine_signal_handler.erl (hunk 7 above)
Comment (zmstone) [reply apps/emqx_machine/src/emqx_machine_signal_handler.erl:39]: it's handled inside of `emqx_machine_terminator:graceful()`

File: apps/emqx_machine/src/emqx_machine.erl (hunk 8 above)
Comment (zmstone) [reply apps/emqx_machine/src/emqx_machine.erl:51]: 1. agreed, will change.
2. one concerns: if `emqx_machine_terminator` is under supervision tree, the call to `init:stop()` will kill itself as a part of `emqx_machine` app, but not *after* `emqx_machine`.

Comment (zmstone) [reply apps/emqx_machine/src/emqx_machine.erl]: a second though: the exist of `emqx_machine` app is probably enough to indicate the node being down since the chance for otp apps to hang is very small.
changed it to reside under supervisor.

**Review by zhongwencool — APPROVED** [review]

**Review by qzhuyan — APPROVED** [review]

**Comment by zmstone** [comment]
Thanks for the review @zhongwencool @qzhuyan
